user/src/com/google/gwt/user/server/rpc/impl/ServerSerializationStreamWriter.java

```diff
@@ -620,14 +503,14 @@ public String toString() {
     // We take a guess at how big to make to buffer to avoid numerous resizes.
     //
     int capacityGuess = 2 * tokenListCharCount + 2 * tokenList.size();
-    LengthConstrainedArray stream = new LengthConstrainedArray(capacityGuess);
-    writePayload(stream);
-    writeStringTable(stream);
-    writeHeader(stream);
+    StringBuffer buffer = new StringBuffer(capacityGuess);
```

Comment:
As long as we're updating this, consider replacing StringBuffer with StringBuilder for better single-threaded performance? From https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/StringBuffer.html

> As of release JDK 5, this class has been supplemented with an equivalent class designed for use by a single thread, [StringBuilder](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/StringBuilder.html). The StringBuilder class should generally be used in preference to this one, as it supports all of the same operations but it is faster, as it performs no synchronization.

Reply:
I updated to use StringBuilder instead of StringBuffer and changed the default serialization format to JSON.

I'm not sure how best to implement the rest of your suggestions though. But they seem logical.